Change to accept NumPy 0D arrays in _f90repr

LiamPattinson opened the pull request:
Hi there! We've been using f90nml in [Pyrokinetics](https://github.com/pyro-kinetics/pyrokinetics) to read/write input files to plasma physics codes, and we ran into an issue where NumPy 0D arrays would propagate unnoticed into a `Namelist` and cause an exception to be thrown on calling `Namelist.write`. I've added a small change to `Namelist._f90repr` to detect 0D arrays and convert them to their equivalent primitive type. Nothing changes if NumPy is not installed.

Let me know if you'd like me to add further test cases, or if you'd prefer to avoid including these sorts of edge cases in your library.

marshallward commented:
I think this patch is addressing the output stage, but the issue is that the 0d array was even allowed into the Namelist.

It is good news that they were able to persist, but it might be a better idea to catch these things as they are instantiated and turn them into zero-length lists.

Would something like that solve your issue?  Or do you still need to preserve the type of these zero-element arrays?

LiamPattinson commented:
That's a fair point, it would make more sense for the `Namelist` class to only allow Fortran-compatible types in the first place. I can't think of any good reason to preserve 0D array types within the namelist -- they usually decay into primitive types after being operated on anyway. I've moved the conversion step to `__setitem__`. Will I need to perform similar checks elsewhere?

marshallward commented:
Thanks for the update, I think addressing this at `__setitem` is probably the better choice.

I wonder if we even need to check if it's a NumPy array.  It seems that any empty container is going to have this issue.  Perhaps its simpler to just convert any empty container to a zero-element list?  That would avoid the need to import numpy.

I am not sure if this is possible, nor what the test would be (`len() == 0`?) and might put unrealistic expectations on classes.  But if it works then it might be a more long term solution.

marshallward commented:
As an aside: Handling of general objects into the `Namelist()` constructor is an overall blind spot in this module.  If I remember, there are many ways to break `Namelist()` with lists of exotic types.

This PR is not necessarily connected to that issue, since numpy arrays are very close to supported types (i.e. lists of numbers), and certainly doesn't need to try and solve it, but might be useful context for the discussion.

LiamPattinson commented:
A worry that I have with this update is that the scope for any similar changes is unbounded. I encountered similar issues with 0D Xarray DataArrays and with quantities with physical units using Pint, but `isinstance(value, np.ndarray)` won't catch either of those cases. It would be unreasonable to try to catch every similar type in the Python ecosystem, not to mention other edge cases such as 'objects which look like a string but actually aren't'. However, a _lot_ of libraries are NumPy aware, so we could try something like:

```python
try:
    # Try converting to numpy array. If an exception is raised, ignore this step
    # Note: may need to specify dtype=object
    as_array = np.asanyarray(value)
    # If resulting object is a 0D array, decay to primitive type
    if np.ndim(as_array) == 0:
        # TODO handle special cases of numpy string/bool types
        value = as_array[()]
except:
    pass
```

This would catch a large amount of scalar-like objects and successfully return the equivalent primitive, but it would also catch everything else that passed through, and I'm not sure if there would be any unexpected side effects or performance issues. It'll take some experimentation for me to figure out if there's some catch-all solution.

> I wonder if we even need to check if it's a NumPy array. It seems that any empty container is going to have this issue. Perhaps its simpler to just convert any empty container to a zero-element list? That would avoid the need to import numpy.

The issue is that a NumPy 0D array isn't empty, but just has no dimensions. I think they only exist because the NumPy developers found it cleaner to have some reduction routines return a 0D array instead of a primitive, presumably because you can still perform operations like `x.shape` on them. They usually decay quietly to primitives when you use them in most expressions, but sometimes they don't, and I can't quite figure out what the rules are.

>I am not sure if this is possible, nor what the test would be (len() == 0?) and might put unrealistic expectations on classes. But if it works then it might be a more long term solution.

[NumPy's `isscalar`](https://numpy.org/doc/stable/reference/generated/numpy.isscalar.html) is a good candidate, as is `np.ndim()==0`. The latter catches 0D array, but also a lot of other things we probably don't want to catch. The former should let things like bools and string through, but not more complex classes. Maybe some combination of the two would work? It would be better to avoid NumPy as a required dependency though.

I'll have a think over the next couple of days to see if I can come up with a better solution to what I have here.

marshallward commented:
> A worry that I have with this update is that the scope for any similar changes is unbounded.

I agree, this is the issue that I am also concerned about.  A solution independent of type will hopefully blunt this problem.

Some type-agnostic options for testing if a field has length:

https://stackoverflow.com/questions/23567628/how-to-check-if-len-is-valid

From my simple tests, NumPy arrays appear to understand `len() == 0` as an empty array.
```
>>> import numpy as np
>>> n = np.array([], dtype=np.float32)
>>> len(n)
0
```
Assuming `len( ) == 0` also works for Xarray and DataArrays, I think this could end up being a good option, and would cover more exotic types.  Nonzero length is hard, but zero-length is unambiguously empty.

I can't think of any cases where `len()` is problematic, but you have thought about this for longer and may know about them.

>  I think they only exist because the NumPy developers found it cleaner to have some reduction routines return a 0D array instead of a primitive.

Fortran understands them too, so no argument here :laughing: .  Though most namelists would struggle to distinguish them.

LiamPattinson commented:
I did some digging, and found that a large number of objects from NumPy-friendly libraries implement the `__array_ufunc__` protocol from [NEP 13](https://numpy.org/neps/nep-0013-ufunc-overrides.html). By checking if an object has this, we can affect many more objects than just `np.ndarray`.

Just to make sure we're on the same page, I've been aiming to solve the issue of 0D `np.ndarray`s of scalars, not empty arrays. These behave quite confusingly:
```python
np.array(5).shape == ()
len(np.array(5)) # Raises exception, 0D array doesn't have a len
np.isscalar(np.array(5)) == False
np.ndim(np.array(5)) == 0
```

This is different to an empty array:
```python
np.array([]).shape == (0,)
len(np.array([])) == 0
np.isscalar(np.array([])) == False
np.ndim(np.array([])) == 1
```

And also different to a NumPy scalar type:
```python
np.float64(5).shape # Raises exception, doesn't have a shape
len(np.float64(5)) # Raises exception, doesn't have a len
np.isscalar(np.float64(5)) == True
np.ndim(np.float64(5)) == 0
```

In my latest attempt, I've written a function `to_primitive` which tries to handle each of these cases, and in my own testing it appears to work for Xarray types too. It doesn't work for units libraries like AstroPy or Pint, and I'm also not sure about things CuPy or Dask.

marshallward commented:
I apologize, this is something different, and I've never seen it before.

I find it odd that something like `np.array(5)` behaves as an integer, interacts with integers without much problem, but then `isinstance(np.array(5), numbers.Integral)` reports as `False`.

Since these things behave like their primitives in most respects, perhaps there is some way to detect this without bringing in the machinery of NumPy.  I would like to look into this more before bringing a solution which depends on NumPy.  It might also help for compatibility with these other projects.

(BTW: f90nml does also fail on zero-sized arrays, but this is a separate issue.)

marshallward commented:
I am looking at the overall compatibility with NumPy, and it is far more problematic than I had thought.

This example, based on the test suite, does work:
```
>>> d = {'grp': {'x': np.int64(123)}}
>>> nml = f90nml.Namelist(d)
>>> print(nml)
&grp
    x = 123
/
```
but the field never loses its NumPy state.
```
>>> type(nml['grp']['x'])
<class 'numpy.int64'>
```
The general compatibility of NumPy data and intrinsics, as well as this check,
```
>>> isinstance(nml['grp']['x'], numbers.Integral)
True
```
preserves something close to compatibility.  But I am surprised that it slipped in without issue.  I probably would not have chosen to do this on purpose.

However, if we do anything more complex, then things quickly fall apart:
```
>> d = {'grp': {'x': np.array([1,2,3])}}
>>> nml = f90nml.Namelist(d)
>>> print(nml)
Traceback (most recent call last):
  ...
  File ".../f90nml/namelist.py", line 970, in _f90repr
    raise ValueError('Type {0} of {1} cannot be converted to a Fortran'
ValueError: Type <class 'numpy.ndarray'> of [1 2 3] cannot be converted to a Fortran type.
```
Similar for the case we already know
```
d = {'grp': {'x': np.array(1)}}
>>> nml = f90nml.Namelist(d)
>>> print(nml)
Traceback (most recent call last):
  ...
  File ".../f90nml/namelist.py", line 970, in _f90repr
    raise ValueError('Type {0} of {1} cannot be converted to a Fortran'
ValueError: Type <class 'numpy.ndarray'> of 1 cannot be converted to a Fortran type.
```
My take here is that the `Namelist()` constructor is unprepared to handle anything from NumPy, not just 0D arrays.

Handling this would be expanding the scope of your issue, but this might be a good chance to address the larger problem.

LiamPattinson commented:
With a little tweak, I think I've got it handling NumPy arrays of one dimension or more too! If we detect a NumPy array (or something similar) with one dimension or more, we convert to a list, and recursively call `to_primitive` on the contents until we're left at a list (of lists (of lists...)) of Python primitive types. I added a catch for empty arrays too.

> perhaps there is some way to detect this without bringing in the machinery of NumPy

I added this part to ensure the user doesn't need to have NumPy installed to use f90nml:

```python
try:
    import numpy as np
    has_numpy = True
except ImportError:
    has_numpy = False
```
If the user already has NumPy installed, `to_primitive` will check for NumPy types (or those from related libraries). If the user doesn't have NumPy, we're not going to encounter any of those types anyway, so we can safely skip the function.

marshallward commented:
I understand that is how it works.  I am more concerned about carving out an exception for NumPy rather than addressing the broader issue of non-intrinsic types, or iterables of such types, being passed through the constructor.

I am wondering if it's better to "duck-type" these, for example checking if the object or its contents behave as integers as the criterion, rather than explicitly relying on either `int` or `np.int64`.

marshallward commented:
FWIW I have been able to make some progress using the `tolist()` method:
```
>>> import numpy as np
>>> a = np.array([1,2,3])
>>> a.tolist()
[1, 2, 3]
>>> b = np.array(1)
>>> b.tolist()
1
>>> c = np.array(1.23)
>>> c.tolist()
1.23
```
Not the greatest solution, since it still relies on existence of `tolist` in the object, but it may avoid the conditional module load.  And the fact that `tolist()` does not return a list seems pretty strong way to detect if something's a scalar.

marshallward commented:
Actually, perhaps it's not just better for the NumPy user to call `tolist()` before passing to f90nml?

LiamPattinson commented:
> FWIW I have been able to make some progress using the `tolist()` method:

I wasn't aware of that one! Oddly, it even works on scalar types, so `np.float64(1.0).tolist()` returns a Python float, but it stops working if precision would be lost, so `np.float128(1.0).tolist()` just returns a `np.float128`. It looks like `pandas.Series` also implements `.tolist()`, but `xarray.DataArray` doesn't (all three work using my `to_primitive` function). I'd argue that it's not worth avoiding `.tolist()` just for Xarray though, as you can trivially extract the underlying NumPy array using `data_array.values` or `data_array.data`, and regular users of Xarray are probably used to doing so quite regularly to get around these sorts of compatibility issues.

>  I am more concerned about carving out an exception for NumPy rather than addressing the broader issue of non-intrinsic types, or iterables of such types, being passed through the constructor.

I'd guess that the majority of people using the library will be using it for science/engineering related purposes, and NumPy is ubiquitous in those spaces, so it could still be helpful without solving the much larger problem of generic non-intrinsic types. The changes I've made go quite a bit further than just NumPy too, as they also work on many other objects within the wider scientific computing ecosystem.

> I am wondering if it's better to "duck-type" these, for example checking if the object or its contents behave as integers as the criterion, rather than explicitly relying on either int or np.int64.

> Actually, perhaps it's not just better for the NumPy user to call tolist() before passing to f90nml?

I feel like this PR is being pulled in a few different directions. Which of these approaches would you prefer I take?

- Coerce NumPy arrays (and similar objects) into Python built-in types within `__setitem__`, so prevent them from getting into the Namelist to begin with. I can get rid of the conditional NumPy import and `to_primitive` at this stage using `.tolist()`.
- Allow NumPy types to propagate as before in a 'duck typing' style, and only handle errors where they pop up (i.e. printing, writing to file), similarly to how I originally tried to catch these problems in `_f90repr`.
- Don't try to catch these errors, and instead just update the docs to warn users about non-intrinsic types and recommend solutions such as `.tolist()`.

I'd personally lean in favour of either of the first two solutions, as users are very likely to be using NumPy and/or its wider ecosystem, and this would minimise friction between their various dependencies. I don't have a strong preference between duck-typing and explicit type checking, as each have benefits and drawbacks.

marshallward commented:
> I feel like this PR is being pulled in a few different directions.

You are not imagining it! :grin: Mostly because I had a lot of misconceptions about the current state of NumPy compatibility:

* I knew that NumPy scalars were supported, but did not realize they were retaining their original types (`float64`, etc.).

* I incorrectly thought that NumPy arrays were supported.  I now know that they are not.

* I had not even known about NumPy 0D arrays, whose unusual behavior further complicates the issue.

This is all layered on the broader unanswered question of how to handle generic objects.

I apologize for the unfocused discussion.  This new information has forced me to keep re-evaluating how this could be handled.

Ideally, I'd like to believe that there is some agnostic ("Pythonic"?) way to traverse an `ndarray` and extract its components with intrinsics or the standard library.  But as yet, I cannot find any way to do that.

Usually `json` is my guiding light in these situations:
```
>>> import numpy as np
>>> import json
>>> a = np.array([1,2,3])
>>> json.dumps([1,2,3])
'[1, 2, 3]'
>>> json.dumps(a)
Traceback (most recent call last):
  ...
  File "/usr/lib/python3.10/json/encoder.py", line 179, in default
    raise TypeError(f'Object of type {o.__class__.__name__} '
TypeError: Object of type ndarray is not JSON serializable
>>>
```
If `json` or other standard data serializers don't know how to serialize a NumPy array, then it's probably not very realistic to expect `Namelist` to solve the problem.  Perhaps even the NumPy team do not expect its `ndarray` to be well-behaved outside of NumPy.

We could force the issue with conditional imports or `hasattr(x, 'tolist')` checks, but I feel like this is motivated by an overlapping userbase rather than good practice.  It seems to me that the user ought to be the one to pull the `ndarray` out of its NumPy space and back into the common Python space.  If that's the case, then I think option 3 is the correct one.

I would be interested to hear your thoughts though.  I really appreciate your willingness to keep up this dialogue, I hope it hasn't been too exhausting!

LiamPattinson commented:
I've pushed a change to using `tolist` in `__setitem__`, so there's now no explicit NumPy dependence at all (it may be worth also checking for `hasattr(value, "__array_ufunc__")` so we don't hit unrelated `tolist` functions that some other objects might implement). That's where I'd prefer to leave this PR, since it looks like it would be difficult to handle generic arrays at the writing stage, and it sounds like this feature is unlikely to be merged in any case.

While I wouldn't expect Python's JSON module to support NumPy, it does provide a way for users to extend its applicability using [`JSONEncoder` subclasses](https://docs.python.org/3/library/json.html#json.JSONEncoder), so the following works:

```python
>>> class NumpyEncoder(json.JSONEncoder):
...     def default(self, obj):
...             if hasattr(obj, "tolist"):
...                     return obj.tolist()
...             return super().default(obj)
... 
>>> json.dumps(np.array([1,2,3]), cls=NumpyEncoder)
'[1, 2, 3]'
```

Perhaps something similar could work here? It would require a fairly serious rewrite, as functions such as `_var_strings` and `_f90repr` would need to be spun off into an 'encoder' object which users could subclass for their own purposes.

The use case that motivated this PR is that we're trying to build fairly complex input files for plasma physics software used in fusion energy research, and we're drawing information from a variety of sources. Since `Namelist` will accept objects without complaint, but later raise exceptions when writing to file, it can be difficult to track down the places where bad data is getting into the `Namelist`, and I found that the easiest solution was to 'monkey patch' `_f90repr` to correct any issues at the point of writing a file. It's not exactly an elegant solution, but it works fairly well, so we can always continue with that for now.

marshallward commented:
This looks like a very concise solution, I think that I'd like to add it.  Perhaps it breaks some of the rules that I've laid out, but it's very localized and can be modified without much trouble if a better solution presents itself.

Your suggestion to include custom encoders is also a great idea, thanks for pointing it out.  I had not known about it.  It's obviously more ambitious and the project would need a major refactor, but it does look like the correct approach in the long term.

Are you OK with a squash merge?  Or did you want to preserve/reformat the commits?

LiamPattinson commented:
Thanks, I'm glad I was able to contribute. I'm happy for it to be a squash merge, as I don't think there's much value in preserving the intermediate solutions.
